packages/labs/ssr-dom-shim/src/lib/events.ts

```diff
@@ -0,0 +1,474 @@
+
+const enumerableProperty: Record<string, unknown> = {__proto__: null};
+enumerableProperty.enumerable = true;
+Object.freeze(enumerableProperty);
+
+// TODO: Remove this when we remove support for vm modules (--experimental-vm-modules).
+const EventShim = class Event implements EventInterface {
+  #cancelable = false;
+  #bubbles = false;
+  #composed = false;
+  #defaultPrevented = false;
+  #timestamp = Date.now();
+  #propagationStopped = false;
+  #type: string;
+  #target: EventTarget | null;
+  #isBeingDispatched: boolean;
+  readonly NONE = NONE;
+  readonly CAPTURING_PHASE = CAPTURING_PHASE;
+  readonly AT_TARGET = AT_TARGET;
+  readonly BUBBLING_PHASE = BUBBLING_PHASE;
+  static readonly NONE = NONE;
+  static readonly CAPTURING_PHASE = CAPTURING_PHASE;
+  static readonly AT_TARGET = AT_TARGET;
+  static readonly BUBBLING_PHASE = BUBBLING_PHASE;
+
+  constructor(type: string, options: EventInit = {}) {
+    if (arguments.length === 0)
+      throw new Error(`The type argument must be specified`);
+    if (typeof options !== 'object' || !options) {
+      throw new Error(`The "options" argument must be an object`);
+    }
+    const {bubbles, cancelable, composed} = options;
+    this.#cancelable = !!cancelable;
+    this.#bubbles = !!bubbles;
+    this.#composed = !!composed;
+
+    this.#type = `${type}`;
+    this.#target = null;
+    this.#isBeingDispatched = false;
+  }
+
+  initEvent(_type: string, _bubbles?: boolean, _cancelable?: boolean): void {
+    throw new Error('Method not implemented.');
+  }
+
+  stopImmediatePropagation() {
+    this.stopPropagation();
+  }
+
+  preventDefault() {
+    this.#defaultPrevented = true;
+  }
+
+  get target(): EventTarget | null {
+    return this.#target;
+  }
+
+  get currentTarget(): EventTarget | null {
+    return this.#target;
+  }
+
+  get srcElement(): EventTarget | null {
+    return this.#target;
+  }
+
+  get type(): string {
+    return this.#type;
+  }
+
+  get cancelable(): boolean {
+    return this.#cancelable;
+  }
+
+  get defaultPrevented(): boolean {
+    return this.#cancelable && this.#defaultPrevented;
+  }
+
+  get timeStamp(): number {
+    return this.#timestamp;
+  }
+
+  composedPath(): EventTarget[] {
+    return this.#isBeingDispatched ? [this.#target!] : [];
+  }
+
+  get returnValue(): boolean {
+    return !this.#cancelable || !this.#defaultPrevented;
+  }
+
+  get bubbles(): boolean {
+    return this.#bubbles;
+  }
+
+  get composed(): boolean {
+    return this.#composed;
+  }
+
+  get eventPhase(): number {
+    return this.#isBeingDispatched ? Event.AT_TARGET : Event.NONE;
+  }
+
+  get cancelBubble(): boolean {
+    return this.#propagationStopped;
+  }
+
+  set cancelBubble(value) {
+    if (value) {
+      this.#propagationStopped = true;
+    }
+  }
+
+  stopPropagation(): void {
+    this.#propagationStopped = true;
+  }
+
+  get isTrusted(): boolean {
+    return false;
+  }
+};
+
+Object.defineProperties(EventShim.prototype, {
```

Comment:
Why doesn't this redefine all the class accessors above?

Reply:
This was adapted from https://github.com/nodejs/node/blob/main/lib/internal/event_target.js
I'm afraid I couldn't tell you the reason.

Reply:
My bad, I haven't used `defineProperties` to modify properties, but apparently [it does so](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/defineProperties) for existing properties.